// Copyright 2019 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package google.api.envoy.http.backend_auth;

import "api/envoy/http/common/base.proto";
import "validate/validate.proto";

message BackendAuthRule {
  // Operation name, also known as selector.
  string operation = 1 [(validate.rules).string.min_bytes = 1];

  // Audience used to create the JWT token sent to the backend.
  // If empty, then no JWT token will be created.
  // https://cloud.google.com/endpoints/docs/openapi/openapi-extensions#jwt_audience_disable_auth
  string jwt_audience = 2 [(validate.rules).string.min_bytes = 1];
}

// When it is configured, the filter will fetch id token from Instance Metadata
// Server.
// https://cloud.google.com/compute/docs/instances/verifying-instance-identity
message ImdsIdTokenInfo {
  // The Instance Metadata Server uri.
  api.envoy.http.common.HttpUri imds_server_uri = 1;
}

// When it is configured, the filter will fetch id token from Google Cloud IAM
// after the access token fetched from Instance Metadata Server.
message IamIdTokenInfo {
  // The Google Cloud IAM server uri.
  api.envoy.http.common.HttpUri iam_uri = 1;

  // Information used to get access token.
  api.envoy.http.common.AccessToken access_token = 3;

  // The Service Account Email.
  string service_account_email = 2;
}

message FilterConfig {
  // A list of backend auth rules.
  repeated BackendAuthRule rules = 1;

  oneof id_token_info {
    // Information used to fetch id token from Instance Metadata Server.
    ImdsIdTokenInfo imds_token = 2;

    // Information used to fetch id token from Google Cloud IAM.
    IamIdTokenInfo iam_token = 3;
  }
}
